plugins/osdn/vnids.go

```diff
@@ -18,14 +20,103 @@ const (
 	AdminVNID = uint(0)
 )
 
+func (oc *OsdnController) GetVNID(name string) (uint, error) {
+	oc.vnidLock.Lock()
+	defer oc.vnidLock.Unlock()
+
+	if id, ok := oc.vnidMap[name]; ok {
+		return id, nil
+	}
+	// In case of error, return some value which is not a valid VNID
+	return MaxVNID + 1, fmt.Errorf("Failed to find NetID for namespace: %s in vnid map", name)
+}
+
+// Nodes asynchronously watch for both NetNamespaces and services
+// NetNamespaces populates vnid map and services/pod-setup depend on vnid map
+// If for some reason, vnid map propagation from master to node is slow
+// and if service/pod-setup tries to lookup vnid map then it may fail.
+// So, use this method to alleviate this problem. This method will
+// retry vnid lookup and NetNamespaces in etcd before giving up.
+func (oc *OsdnController) WaitAndGetVNID(name string) (uint, error) {
+	// Try few times up to 2 seconds
+	retries := 20
+	retryInterval := 100 * time.Millisecond
+	for i := 0; i < retries; i++ {
+		if id, err := oc.GetVNID(name); err != nil {
+			return id, nil
+		}
+		time.Sleep(retryInterval)
+	}
+
+	// Check if we can get ID from NetNamespace object
+	netns, err := oc.Registry.GetNetNamespace(name)
+	if err == nil {
+		oc.setVNID(name, netns.NetID)
```

Comment:
Also this seems a bit odd; this function gets called from watchServices() and plugin.go::getVNID() only.  It feels like it shouldn't need to update the VNID map at all, since shouldn't that be done already on startup by populateVNIDMap() or at runtime by watchNetNamespaces() and watchNamespaces()-via-assignVNID()?

I feel like this function should only get what's already in the VNID map (possibly waiting for it).  If the ID isn't in the VNID map yet, it'll get added shortly by watchNetNamespaces() or watchNamespaces().  I think centralizing setting the VNID map is less confusing.